man/im_histplot.3
.TH IM_HISTPLOT 3 "10 May 1991"
.SH NAME
im_histplot \- graph an image of one horizontal or vertical line
.SH SYNOPSIS
#include <vips/vips.h>

int im_histplot(in, out)
.br
IMAGE *in, *out;

.SH DESCRIPTION
.B im_histplot(3) 
plots a graph of a horizontal or vertical image file with
one line. It is suitable for displaying the output of histogram functions
such as im_histgr(3) and for displaying line profiles.

Note that if you try to directly plot the result of 
.B im_histgr(3)
you will often get a very, very large image, since
.B im_histplot(3) 
will draw an image as high (or as wide) as the largest pixel
value in the image (potentially huge). Use
.B im_histnorm(3)
to normalise the histogram maximum before plotting.

The input image should be either one horizontal line (Ysize=1) or
one vertical line (Xsize=1).  If the image is FMTUCHAR, the graph is drawn 
as 256 by line-length pixels. For all other unsigned integer types, 
.B im_histplot(3)
finds the image maximum, and draws the graph as max by line-length pixels.

For signed integer types, 
.B im_histplot(3)
finds minimum and maximum, moves min up to zero, and 
draws the graph as min + max by line-lenth pixels.

For float types, 
.B im_histplot(3)
finds minimum and maximum, and scales the image range so as to make  the
graph square. 

.SH RETURNED VALUE
The function returns 0 on success and -1 on error.
.SH SEE ALSO
im_histgr(3), im_histnorm(3), im_extract(3).
.SH COPYRIGHT
National Gallery and Birkbeck College, 1995


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/histograms_lut/im_histplot.c
/* @(#) im_histplot: plot a 1xany or anyx1 image file as a max x any or 
 * @(#) any x max graph using these rules:
 * @(#)
 * @(#) - unsigned char
 * @(#) 	always output 256 
 * @(#) - other unsigned integer types
 * @(#) 	output 0 - max
 * @(#) - signed int types
 * @(#) 	min moved to 0, max moved to max + min.
 * @(#) - float types
 * @(#) 	min moved to 0, max moved to any (square output)
 * @(#)
 * @(#) usage:
 * @(#)
 * @(#)  	int 
 * @(#)  	im_histplot( hist, histplot )
 * @(#)  	IMAGE *hist, *histplot;
 * @(#) 
 * @(#) Returns non-zero on error
 *
 * Copyright: 1990, N. Dessipris.
 *
 * Author: Nicos Dessipris.
 * Written on: 09/07/1990
 * Modified on : 12/03/1991
 * 20/6/95 JC
 *	- rules rationalised
 *	- im_lineprof removed
 *	- rewritten
 * 13/8/99 JC
 *	- rewritten again for partial, rules redone
 * 19/9/99 JC
 *	- oooops, broken for >1 band
 * 26/9/99 JC
 *	- oooops, graph float was wrong
 * 17/11/99 JC
 *	- oops, failed for all 0's histogram 
 * 14/12/05
 * 	- redone plot function in C, also use incheck() to cache calcs
 * 	- much, much faster!
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Normalise an image using the rules noted above.
 */
static int
normalise( IMAGE *in, IMAGE *out )
{
	IMAGE *t1 = im_open_local( out, "im_histplot:2", "p" );
	double min, max;

	if( in->Coding != IM_CODING_NONE ) {
		im_error( "im_histplot", _( "uncoded only" ) );
		return( -1 );
	}
	if( im_iscomplex( in ) ) {
		im_error( "im_histplot", _( "non-complex only" ) );
		return( -1 );
	}

	if( im_isuint( in ) ) {
		/* Trivial case.
		 */
		if( im_copy( in, out ) )
			return( -1 );
	}
	else if( im_isint( in ) ) {
		/* Move min up to 0. incheck(), because we have to min() so we
		 * might as well save the calcs.
		 */
		if( !t1 || 
			im_incheck( in ) ||
			im_min( in, &max ) ||
			im_lintra( 1.0, in, -min, t1 ) )
			return( -1 );
	}
	else {
		/* Float image: scale min--max to 0--any. Output square
		 * graph.
		 */
		int any;

		if( in->Xsize == 1 )
			any = in->Ysize;
		else
			any = in->Xsize;

		/* incheck(), because we have to min()/max() so we
		 * might as well save the calcs.
		 */
		if( !t1 || 
			im_incheck( in ) ||
			im_min( in, &min ) || 
			im_max( in, &max ) ||
			im_lintra( any / (max - min), in, 
				-min * any / (max - min), out ) )
			return( -1 );
	}

	return( 0 );
}

#define VERT( TYPE ) { \
	TYPE *p1 = (TYPE *) p; \
	\
	for( x = le; x < ri; x++ ) { \
		for( b = 0; b < nb; b++ )  \
			q[b] = p1[b] < x ? 0 : 255; \
		\
		q += nb; \
	} \
}


/* Generate function.
 */
static int
make_vert_gen( REGION *or, void *dummy, IMAGE *in )
{
	Rect *r = &or->valid;
	int le = r->left;
	int to = r->top;
	int ri = IM_RECT_RIGHT( r );
	int bo = IM_RECT_BOTTOM( r );
	int nb = in->Bands;

	int x, y, b;

	for( y = to; y < bo; y++ ) {
		unsigned char *q = (unsigned char *) 
			IM_REGION_ADDR( or, le, y );
		unsigned char *p = (unsigned char *) 
			IM_IMAGE_ADDR( in, 0, y );

		switch( in->BandFmt ) {
		case IM_BANDFMT_UCHAR: 	VERT( unsigned char ); break;
		case IM_BANDFMT_CHAR: 	VERT( signed char ); break; 
		case IM_BANDFMT_USHORT: VERT( unsigned short ); break; 
		case IM_BANDFMT_SHORT: 	VERT( signed short ); break; 
		case IM_BANDFMT_UINT: 	VERT( unsigned int ); break; 
		case IM_BANDFMT_INT: 	VERT( signed int );  break; 
		case IM_BANDFMT_FLOAT: 	VERT( float ); break; 
		case IM_BANDFMT_DOUBLE:	VERT( double ); break; 

		default:
			im_error( "im_histplot", _( "internal error #8255" ) );
			return( -1 );
		}
	}

	return( 0 );
}

#define HORZ( TYPE ) { \
	TYPE *p1 = (TYPE *) p; \
	\
	for( y = to; y < bo; y++ ) { \
		for( b = 0; b < nb; b++ )  \
			q[b] = p1[b] < (ht - y) ? 0 : 255; \
		\
		q += lsk; \
	} \
}


/* Generate function.
 */
static int
make_horz_gen( REGION *or, void *dummy, IMAGE *in )
{
	Rect *r = &or->valid;
	int le = r->left;
	int to = r->top;
	int ri = IM_RECT_RIGHT( r );
	int bo = IM_RECT_BOTTOM( r );
	int nb = in->Bands;
	int lsk = IM_REGION_LSKIP( or );
	int ht = or->im->Ysize;

	int x, y, b;

	for( x = le; x < ri; x++ ) {
		unsigned char *q = (unsigned char *) 
			IM_REGION_ADDR( or, x, to );
		unsigned char *p = (unsigned char *) 
			IM_IMAGE_ADDR( in, x, 0 );

		switch( in->BandFmt ) {
		case IM_BANDFMT_UCHAR: 	HORZ( unsigned char ); break;
		case IM_BANDFMT_CHAR: 	HORZ( signed char ); break; 
		case IM_BANDFMT_USHORT: HORZ( unsigned short ); break; 
		case IM_BANDFMT_SHORT: 	HORZ( signed short ); break; 
		case IM_BANDFMT_UINT: 	HORZ( unsigned int ); break; 
		case IM_BANDFMT_INT: 	HORZ( signed int );  break; 
		case IM_BANDFMT_FLOAT: 	HORZ( float ); break; 
		case IM_BANDFMT_DOUBLE:	HORZ( double ); break; 

		default:
			im_error( "im_histplot", _( "internal error #8255" ) );
			return( -1 );
		}
	}

	return( 0 );
}

/* Plot image.
 */
static int
plot( IMAGE *in, IMAGE *out )
{
	IMAGE *t[5];
	double max;
	int tsize;
	int xsize;
	int ysize;

	if( im_incheck( in ) ||
		im_poutcheck( out ) )
		return( -1 );

	/* Find range we will plot.
	 */
	if( im_open_local_array( out, t, 5, "im_histplot", "p" ) ||
		im_max( in, &max ) )
		return( -1 );
	if( max < 0 ) {
		im_errormsg( "im_histplot: internal error #8254" );
		return( -1 );
	}
	if( in->BandFmt == IM_BANDFMT_UCHAR )
		tsize = 256;
	else
		tsize = ceil( max );

	/* Make sure we don't make a zero height image.
	 */
	if( tsize == 0 )
		tsize = 1;

	if( in->Xsize == 1 ) {
		/* Vertical graph.
		 */
		xsize = tsize;
		ysize = in->Ysize;
	}
	else {
		/* Horizontal graph.
		 */
		xsize = in->Xsize;
		ysize = tsize;
	}

	/* Set image.
	 */
	im_initdesc( out, xsize, ysize, in->Bands, 
		IM_BBITS_BYTE, IM_BANDFMT_UCHAR, 
		IM_CODING_NONE, IM_TYPE_HISTOGRAM, 1.0, 1.0, 0, 0 );

	/* Set hints - ANY is ok with us.
	 */
	if( im_demand_hint( out, IM_ANY, NULL ) )
		return( -1 );
	
	/* Generate image.
	 */
	if( in->Xsize == 1 ) {
		if( im_generate( out, NULL, make_vert_gen, NULL, in, NULL ) )
			return( -1 );
	}
	else {
		if( im_generate( out, NULL, make_horz_gen, NULL, in, NULL ) )
			return( -1 );
	}

	return( 0 );
}

int 
im_histplot( IMAGE *hist, IMAGE *histplot )
{
	IMAGE *norm = im_open_local( histplot, "im_histplot:1", "p" );

	if( !norm )
		return( -1 );
	if( hist->Xsize != 1 && hist->Ysize != 1 ) {
		im_error( "im_histplot", _( "Xsize or Ysize not 1" ) );
		return( -1 );
	}

	if( normalise( hist, norm ) ||
		plot( norm, histplot ) )
		return( -1 );
	
	return( 0 );
}
